File: apps/mobile/lib/ai/orchestrator/prompts.ts
import { z } from 'zod';

// ─── System Prompts ──────────────────────────────────────────────────────────

export const TRANSACTION_EXTRACTION_PROMPT = `You are TransactionExtractionAgent for a personal finance app.

You receive user input describing a financial transaction. Extract the following fields:

Required:
- amount: a positive number (the transaction amount)

Best-effort (infer if possible, null if not):
- type: "income" or "expense" (default to "expense" if unclear)
- currency: 3-letter ISO code (default "MYR" if unclear)
- merchant: business/person/app name
- description: short summary of what the transaction is
- wallet_hint: any mention of which account/wallet/card was used
- category_hint: spending category (e.g. "food", "transport", "entertainment")

Rules:
1. Return ONLY valid JSON matching the schema.
2. If the user says "spent", "paid", "bought" → type is "expense".
3. If the user says "received", "earned", "got paid" → type is "income".
4. Extract the exact numerical amount. If multiple amounts exist, pick the most likely transaction amount.
5. If you truly cannot determine an amount, set amount to null.
`;

export const RECEIPT_EXTRACTION_PROMPT = `You are ReceiptExtractionAgent for a personal finance app.

You receive an image of a receipt or financial document. Analyze the image and extract:

Required:
- amount: the total/final amount (positive number)

Best-effort (infer if possible, null if not):
- type: "income" or "expense" (receipts are usually "expense")
- currency: 3-letter ISO code
- merchant: the store/business name from the receipt
- description: brief summary of the purchase
- wallet_hint: payment method if visible (e.g. "Visa ending 4242")
- category_hint: spending category

Rules:
1. Return ONLY valid JSON matching the schema.
2. Prefer the final amount paid (including tax) — "Grand total", "Total due", "Payment amount", "Amount paid" — over subtotal before tax.
3. The merchant name is usually at the top of the receipt.
4. Default type to "expense" for receipts.
`;

/** Sub-agent 1 (native vision): totals only — small JSON, focused inference. */
export const RECEIPT_AMOUNT_VISION_PROMPT = `You are ReceiptAmountAgent for a personal finance app.

Find the ONE amount the customer actually paid or owes for this purchase (the cashflow amount to log).

Return ONLY a single JSON object (no prose, no markdown):
{"amount": number|null, "type": "income"|"expense"|null, "currency": string|null}

Priority for amount (pick the first that clearly applies):
1. Lines such as "Payment amount", "Amount paid", "Total paid", "Grand total", "Total due", "Balance due", "NETT TOTAL", "AMOUNT PAYABLE" — these usually include tax and service charges.
2. If both "subtotal" (before tax) and a larger "total" / "grand total" / tax-inclusive total appear, use the TAX-INCLUSIVE / FINAL total the customer paid — NOT the pre-tax subtotal alone.
3. If only a subtotal is visible with no total, use that subtotal.
4. Ignore unrelated numbers (change given, loyalty points, unit prices without a clear line total).

Rules:
- amount = one positive number, or null if truly unreadable.
- type = almost always "expense" for purchases; "income" only for clear refunds/credits to the customer.
- currency = 3-letter ISO code if visible (MYR, USD, …); else null.
`;

/** Sub-agent 2 (native vision): merchant, narrative fields, payment hints from image + user note. */
export const RECEIPT_DETAILS_VISION_PROMPT = `You are ReceiptDetailsAgent for a personal finance app.

Read the receipt image carefully and produce rich, specific text for the user to recognize this transaction later.

Return ONLY a single JSON object (no prose before or after, no markdown):
{"merchant": string|null, "description": string|null, "wallet_hint": string|null, "category_hint": string|null}

Fields:
1. merchant — Store or business name as printed (header/logo); include branch or mall if printed and helpful.
2. description — 2–4 sentences when possible: what category of purchase this is, main items or departments (e.g. "groceries and household", "dinner for two", "fuel"), quantities or notable lines if readable, time of day or meal if implied. Avoid generic one-word answers like "purchase" or "items" unless nothing else is visible.
3. wallet_hint — Payment method printed on the receipt (card type, last 4 digits, e-wallet, cash). If a user message is provided below and the receipt does not contradict it, include how they said they paid so it can match a named wallet (e.g. user says "cash" → include "cash").
4. category_hint — Short label: food, groceries, transport, health, entertainment, utilities, etc., or null.

Do not leave description empty if any line items, department names, or totals labels are visible — summarize them.
`;

export const WALLET_RESOLUTION_PROMPT = `You are WalletResolutionAgent.

Goal: Match a transaction to one of the user's wallets.

Rules:
1) Review the available wallets provided in the prompt.
2) Compare the wallet_hint/source with wallet names.
3) If a user message states how they paid (cash, debit, a bank/card name), prefer a wallet whose name matches that payment method or account.
4) Return JSON with fields: action, walletId, reason.
5) action must be "create" (confident match) or "skip" (no match).
6) If there is only one wallet, use that wallet.
7) Never invent walletId values — only use IDs from the provided wallet list.
`;

// ─── Schemas ─────────────────────────────────────────────────────────────────

export const extractionResultSchema = z.object({
  amount: z.number().positive().nullable(),
  type: z.enum(['income', 'expense']).nullable().optional(),
  currency: z.string().nullable().optional(),
  merchant: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
  wallet_hint: z.string().nullable().optional(),
  category_hint: z.string().nullable().optional(),
});

/** Native vision — amount sub-agent output */
export const receiptAmountVisionSchema = z.object({
  amount: z.number().positive().nullable(),
  type: z.enum(['income', 'expense']).nullable().optional(),
  currency: z.string().nullable().optional(),
});

/** Native vision — details sub-agent output */
export const receiptDetailsVisionSchema = z.object({
  merchant: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
  wallet_hint: z.string().nullable().optional(),
  category_hint: z.string().nullable().optional(),
});

export const walletDecisionSchema = z.object({
  action: z.enum(['create', 'skip']),
  walletId: z.string().nullable().optional(),
  reason: z.string(),
});

File: apps/backend/internal/extract/prompts.go
package extract

// Prompts ported from the former on-device pipeline (see docs/AI.md),
// extended with confidence/reasoning fields so the backend returns a
// complete Extraction in one call.

// walletSelectionRules is appended to every extraction system prompt. The
// user's wallet list is injected into the user message at request time.
const walletSelectionRules = `

Wallet selection:
- The user message includes AVAILABLE_WALLETS: a JSON array of {id, name, type?, currency?, accountHint?}.
- Set wallet_id to the id of the wallet this transaction belongs to (source / "from" for transfers).
- For transfers only, set transfer_to_wallet_id to the destination wallet id ("to").
- Only use ids from AVAILABLE_WALLETS — never invent wallet ids.
- Match by wallet name, type, payment method, or accountHint when multiple wallets share the same banking app.
- For transfers, cross-check names AND type fields (e.g. type "cash" vs type "bank") when the user names only one side.
- If only one wallet is listed, use its id unless the input clearly refers to a different account.
- If uncertain, set wallet_id to null and set wallet_hint with your best textual guess.`

// transferDetectionRules applies to text extraction only. Transfers are
// between the user's own wallets — use AVAILABLE_WALLETS to infer both sides.
const transferDetectionRules = `

Transfer detection (between the user's OWN wallets in AVAILABLE_WALLETS):
- A transfer moves money between wallets the user owns. No external merchant or person is the final recipient.
- Use type "transfer" when the input implies money moved from one owned wallet to another — even if the user never says "transfer".
- Paying a shop, person, bill, or merchant is NEVER a transfer — use "expense".
- Receiving salary, refunds, or gifts from others is NEVER a transfer — use "income".

Infer transfers from everyday language + AVAILABLE_WALLETS:
- Deposit / top-up INTO a bank, savings, or e-wallet: money came FROM a more liquid wallet (usually cash or another account).
  Examples: "deposited 100 cash to bank", "topped up Maybank", "put 50 into savings" -> wallet_id = source (often cash), transfer_to_wallet_id = destination (bank/savings/e-wallet).
- Withdrawal OUT OF bank/ATM INTO cash or on-hand: money left a bank account and became cash.
  Examples: "withdrew 200 from ATM", "took cash out of Maybank" -> wallet_id = bank, transfer_to_wallet_id = cash/on-hand wallet.
- Explicit moves: "transfer/moved/shifted X from A to B", "from savings to checking".

Direction when only one side is named:
- "deposited … to bank" with both Cash and Bank wallets -> infer Cash -> Bank (user must already hold the cash to deposit it).
- "withdrew … to cash" with Cash and Bank wallets -> infer Bank -> Cash.
- "topped up TNG from Maybank" -> Maybank -> TNG eWallet.
- Use wallet type (cash, bank, savings, credit, e-wallet) to disambiguate when names are generic ("bank", "cash").

For transfers: merchant = null, category_hint = null, set wallet_id + transfer_to_wallet_id when confident (hints only when unsure).`

const textExtractionPrompt = `You are TransactionExtractionAgent for a personal finance app.

You receive user input describing a financial transaction. Extract the following fields:

Required:
- amount: a positive number (the transaction amount)

Best-effort (infer if possible, null if not):
- type: "income", "expense", or "transfer"
- currency: 3-letter ISO code (default "MYR" if unclear)
- merchant: business/person/app name (null for transfers between own accounts)
- description: short summary of what the transaction is
- wallet_id: id from AVAILABLE_WALLETS for the source wallet (expenses, income, transfers)
- wallet_hint: optional text label when wallet_id is null
- transfer_to_wallet_id: destination wallet id for transfers only
- transfer_to_wallet_hint: optional text label when transfer_to_wallet_id is null
- category_hint: spending category (e.g. "food", "transport") — null for transfers
- confidence: number 0-1, how confident you are this is a real transaction with correct fields
- reasoning: one short sentence explaining your extraction

Rules:
1. Return ONLY valid JSON with keys: amount, type, currency, merchant, description, wallet_id, wallet_hint, transfer_to_wallet_id, transfer_to_wallet_hint, category_hint, confidence, reasoning.
2. If the user says "spent", "paid", "bought", or sent money to a person/merchant -> type is "expense".
3. If the user says "received", "earned", "got paid" from an external payer -> type is "income".
4. If the user moves money between their own accounts -> type is "transfer" and set wallet_id + transfer_to_wallet_id (or hints when unsure). This includes deposits, withdrawals, top-ups, and explicit "from X to Y" moves — see transfer rules below.
5. Paying another person or external merchant is NEVER a transfer — use "expense".
6. Depositing your own cash into your own bank account is a transfer (cash -> bank), NOT income.
7. Withdrawing from your bank to your own cash is a transfer (bank -> cash), NOT income.
8. Extract the exact numerical amount. If multiple amounts exist, pick the most likely transaction amount.
9. If you truly cannot determine an amount, set amount to null.` + transferDetectionRules + walletSelectionRules

const receiptExtractionPrompt = `You are ReceiptExtractionAgent for a personal finance app.

You receive an image of a receipt or financial document. Analyze the image and extract:

Required:
- amount: the total/final amount (positive number)

Best-effort (infer if possible, null if not):
- type: "income" or "expense" (receipts are usually "expense")
- currency: 3-letter ISO code
- merchant: the store/business name from the receipt
- description: 2-4 sentences: what category of purchase this is, main items or departments, quantities or notable lines if readable. Avoid generic one-word answers.
- wallet_id: id from AVAILABLE_WALLETS for how the user paid
- wallet_hint: payment method text when wallet_id is null (e.g. "Visa ending 4242")
- category_hint: spending category (food, groceries, transport, health, entertainment, utilities, etc.)
- confidence: number 0-1
- reasoning: one short sentence explaining what you read from the receipt

Rules:
1. Return ONLY valid JSON with keys: amount, type, currency, merchant, description, wallet_id, wallet_hint, category_hint, confidence, reasoning.
2. Prefer the final amount paid (including tax) — "Grand total", "Total due", "Payment amount", "Amount paid", "NETT TOTAL", "AMOUNT PAYABLE" — over subtotal before tax.
3. If both a subtotal and a larger tax-inclusive total appear, use the TAX-INCLUSIVE / FINAL total — NOT the pre-tax subtotal.
4. The merchant name is usually at the top of the receipt.
5. Default type to "expense" for receipts.
6. If a user message states how they paid and the receipt does not contradict it, pick the matching wallet_id from AVAILABLE_WALLETS.
7. Ignore unrelated numbers (change given, loyalty points, unit prices without a clear line total).` + walletSelectionRules

const notificationDetectionPrompt = `You are a strict notification transaction detector for a personal finance app.

Your task:
1) Decide if a notification is a real financial transaction.
2) Only classify as transaction when all are true:
   - Source app is a bank, fintech, payment, or wallet app.
   - Message contains a real money amount.
   - Message indicates money movement to or from a person/business/merchant.

Treat as NOT a transaction:
- Promotions, ads, cashback campaigns, coupons, reminders.
- OTP/security alerts/login/device alerts.
- Generic balance snapshots without a transaction event.
- Bills due notices without confirmed payment.

If transaction=true:
- Extract amount as a positive number.
- Infer type:
  - income: credited/received/refund inbound.
  - expense: debited/paid/spent/purchase outbound.
- Currency must be 3-letter ISO when possible (USD, MYR, INR, etc).
- Merchant/counterparty should be null when unknown.
- Set wallet_id from AVAILABLE_WALLETS when the notification app or text clearly maps to one wallet (e.g. Maybank app -> Maybank wallet). When multiple wallets share the same app, use accountHint and notification body (account name, last digits, product type) to pick the correct wallet_id. Otherwise null and optional wallet_hint.

Return ONLY valid JSON. When not a transaction:
{"is_transaction": false, "reasoning": "..."}
When a transaction:
{"is_transaction": true, "reasoning": "...", "confidence": 0.0, "amount": 0, "currency": "MYR", "type": "income|expense", "merchant": null, "description": null, "wallet_id": null, "wallet_hint": null, "category_hint": null}` + walletSelectionRules
